e2e-examples/gcs/benchmark/channel_poller.cc

```diff
@@ -0,0 +1,39 @@
+#include "channel_poller.h"
+
+#include <chrono>
+
+ChannelPoller::ChannelPoller(std::shared_ptr<grpc::Channel> channel)
+    : channel_(channel) {
+  StartWatch();
+  thread_ = std::unique_ptr<std::thread>(
+      new std::thread([this]() { this->ThreadRun(); }));
+}
+
+ChannelPoller::~ChannelPoller() {
+  {
+    absl::MutexLock lock(&mu_);
+    shutdown_ = true;
+    cq_.Shutdown();
+  }
+  thread_->join();
+}
+
+void ChannelPoller::StartWatch() {
+  grpc_connectivity_state last_observed = channel_->GetState(true);
+  channel_->NotifyOnStateChange(
+      last_observed, std::chrono::system_clock::now() + std::chrono::seconds(1),
```

Comment:
nit: actually we can use an infinite timeout here (I should have done that in my example)

Reply:
Shutdown appears to be delayed related to this timeout. The more timeout it has, the longer `join` after calling `shutdown` takes because cq.Next() waits. I'm going to follow up on this after getting this merged first.

Reply:
In order to terminate the loop abruptly, we need to fully destroy the channel during `ChannelPoller`'s dtor. That will give us a connectivity state change to SHUTDOWN

Reply:
Oh I got it. https://github.com/GoogleCloudPlatform/grpc-gcp-cpp/pull/34